## Project Data

**File Type:** fastq

**RADseq run:** Illumina HiSeq4000, Paired end reads, 150bp

In [1]:
pwd

u'C:\\Users\\Mary Fisher\\Documents\\Github'

In [2]:
!which bash

/usr/bin/bash


In [3]:
cd ../../../../

C:\


In [4]:
cd ../

C:\


In [5]:
cd D:/

D:\


In [6]:
cd Pacific cod

D:\Pacific cod


In [7]:
cd DataAnalysis/raw_data/L1_PE150

D:\Pacific cod\DataAnalysis\raw_data\L1_PE150


#### Raw Data Files from Lane 1 RADseq

In [9]:
ls

 Volume in drive D is MARY
 Volume Serial Number is 58D4-FF55

 Directory of D:\Pacific cod\DataAnalysis\raw_data\L1_PE150

10/17/2016  01:17 PM    <DIR>          .
10/17/2016  01:17 PM    <DIR>          ..
10/14/2016  04:30 PM    <DIR>          .ipynb_checkpoints
10/07/2016  03:19 PM    23,541,160,542 643_S1_L001_R1_001.fastq.gz
10/17/2016  10:52 AM    90,192,439,385 643_S1_L001_R2_001.fastq
10/07/2016  03:23 PM    27,857,684,012 643_S1_L001_R2_001.fastq.gz
10/10/2016  08:55 AM    <DIR>          FastQC
10/17/2016  09:40 AM    <DIR>          test
10/17/2016  09:41 AM    <DIR>          test2
               3 File(s) 141,591,283,939 bytes
               6 Dir(s)  1,775,118,778,368 bytes free


*The two fastq.gz files are the original, zipped raw data downloaded from OSU's website*

In [10]:
!which bash #have been having some trouble executing code; want to make sure 'bash' is on the machine

/usr/bin/bash


### Taking a look at the raw data...


(1) Forward Sequences

In [11]:
%%bash 
gunzip -c 643_S1_L001_R1_001.fastq.gz | head -n 20

@K00337:24:HCJGYBBXX:1:1101:1641:1173 1:N:0:1
NCAGCGTGCAGGAGGACACCCAGTACATGGCCATGAGGCAGCAGCCCGTCCACCAGAAGCAGAGGTCAGTGTGCTCTCTCCCCCCCCAAAGGCGTTCCCCCCTGTGCGGCCGCGCAGCTGACACAATCTCATCCAGGTTGTATTCCGTTCA
+
#AAFFJ-777-77FAJJFJJJJJJJJJJJJJJJJJFJJJJFJJFJJJJJJJFJJJJJJFJJJJJF-FFJAFJJFJAFJFJ---7---777--A--7---7-7---7--7-A))))-))-A7)-------77----7))7---7------)-
@K00337:24:HCJGYBBXX:1:1101:1722:1173 1:N:0:1
NATCGCAGTTCGCTACACGCAGGACGCTTTTTCACGTTCTGGTTGGTTGTGGCCTGTTGATGCTAAAGGTGAGCCGCTTAAAGCTACCAGTTATATGGCTGTTGGTTTCTATGTGGCTAAATACGTTAACAAAAAGTCAGATATGGACCTT
+
#AAF<FJ-7FJ-AFJJJJJJJAJJFJ<<JJFJJJJJF<JJFJJAFAAF<AJ7JJ<<AAAJJJJFFFJFJJAFFJJJJ<JJJJJJJJJFAFFAFJJJA7FF7FJJFJJFFF7--7F-F<AJJJFA-<FAFFF<FAJ<<A-<FFJ<FAAAJAF
@K00337:24:HCJGYBBXX:1:1101:2067:1173 1:N:0:1
NTTCCGTCACACACACACACACACACACACACACACACACACACACACACACATACACACACACACACACACACACACACACACACACACACACACACACACACACACACACACACACACACACACACACACACCCACACACACACACAAA
+
#AAFFJ<7-AJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJ-7FFJJ<JFJJ-FJJJJJJJJJJJJJJJJJJJ<JJJJFJJAJJ7AJJFJ7FAFF

(2) Reverse Sequences

In [12]:
%%bash 
gunzip -c 643_S1_L001_R2_001.fastq.gz | head -n 20

@K00337:24:HCJGYBBXX:1:1101:1641:1173 2:N:0:1
CCTAGTGTGAACCACCGACACAACAGCAGCACTCCCAGTGAAGTGCAAGGCATAACAATTCTTGAACCCAGAACTAAAATAACAAAAAAAGATAACAAAATCAACCGGAATGCCGATACAAGACCCAGTCTCACCCCGTCGCCAGAACCCC
+
A-7<<-F-<-<A--A-7-F-7--FF-AJ----F-----7AF-7--7<A-7--<A-AA-7-<-------7A-7--7<A7--7-77--7---7A-7-----7----7-)))7----)7)-----777)))))-7<-)7A))))-))-)-F<)-
@K00337:24:HCJGYBBXX:1:1101:1722:1173 2:N:0:1
AATTTCATCCTGAAGTTGCGGCTCATTATGATTCAGAACATTGTCGTGGGAAGTAGCGACAGCTTGGCATTTAGTGTGATGGTCCAATCTCTAGCTCCTATACCTCTAGCAGCAAGTTCCATATCTTACTTTTTGATGAATGAGTAACCCA
+
A-A<7FFJ-F------<----7F<7---F7--<7<-7AAJ----<---7-7---<--7--7--7-<77---7-A-7--7AAJAJ777----77--77F)7---<---AA-<7AA----7------------------7<---------<--
@K00337:24:HCJGYBBXX:1:1101:2067:1173 2:N:0:1
GGCATATTTTTTTGTTGTTGTTTTAGCAGCGAATTGCACCGTGTGCGTGTGTGTGTGTGTGTGTGTGTGCGTGTGTGTGTGTGTGTGTGTGTGTGTGTGTGTGGGTGTGTGTGAGTGGTTGTGCGCGTGTGTGATGTGGGGGAGGGGAGGG
+
AAAFAJFJJ<FA-<FJA---<AFJF<JF-7-AA-A7AAJJF<-77<7--7FAJF<---7-<<7777AFF-<<J<<F--7AFFF<--7A-7A-----

**Format of the Fastq Files:**

*Line 1* = header ("@")

*Line 2* = sequences

*Line 3* = notes (+) - there are not additional notes / repeated header lines in these fastq files

*Line 4* = quality scores (#). HiSeq4000 uses phred 33 scoring

The way the data is structured, all of the forward sequences and all of the reverse sequences are put into a file each. I'll have to run a program in the stacks pipeline process_radtags to demultiplex all of the data into individuals

### How many sequences per file?

Since each sequence has a header that starts with "@", I can count the number of sequences in the file by extracting only the lines that being with "@", and then piping the output to the line count command. 

**!!** As noted in the textbook (p. 342-343), quality scores (Line 4) can sometimes be coded using "@" in fastq files. This means that the sequence counts from the following commands may be an overestimate. A more robust way to count sequences is using `bioawk`. However, I will start with `zgrep` for now. An easier way would also be to wait until the data has been through process_radtags, which puts out a Fasta file for each individual in the data set. I could then get individual-specific sequence counts. 

In [13]:
%%bash
zgrep -c "^@" *.fastq.gz | wc -l >> seq_counts.txt

*This ended up taking a very long time to run...and still had no output after 30 min.*

So I downloaded [gawk](https://sourceforge.net/projects/gnuwin32/files/gawk/3.1.6-1/) (gawk-3.1.6-1-setup.exe) and used the following pipeline from [StephenTurnerGithub](https://github.com/stephenturner/oneliners/blob/40943e6cd4695f3c4d8a1b8d5e940b957e060c26/README.md#awk--sed-for-bioinformatics).
But I entered `zcat` instead of `cat` so that I don't have to unzip my data files. 

According to his **awk & sed for bioinformatics** section on Github, This prints out basic sequence statistics including total number of reads, total number of unique reads, percentage of unique reads, most abundant sequence, its frequency, and percentage of total in file.fq

In [16]:
%%bash
which awk

/usr/bin/awk


In [17]:
%%bash
which zcat

/usr/bin/zcat


In [19]:
%%bash
zcat 643_S1_L001_R1_001.fastq.gz | awk '((NR-2)%4==0){read=$1;total++;count[read]++}END{for(read in count){if(!max||count[read]>max) {max=count[read];maxRead=read};if(count[read]==1){unique++}};print total,unique,unique*100/total,maxRead,count[maxRead],count[maxRead]*100/total}'

266452801 53968105 20.2543 CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC 82415 0.0309304


*Total number of reads:* 266,452,801

*Total number of unique reads:* 53,968,105

In [8]:
%%bash
zcat 643_S1_L001_R2_001.fastq.gz | awk '((NR-2)%4==0){read=$1;total++;count[read]++}END{for(read in count){if(!max||count[read]>max) {max=count[read];maxRead=read};if(count[read]==1){unique++}};print total,unique,unique*100/total,maxRead,count[maxRead],count[maxRead]*100/total}'


gzip: stdout: File too large
awk: cmd. line:1: (FILENAME=- FNR=1021079364) fatal: str_array.c:387:str_list: list: can't allocate 1091045232 bytes of memory (Cannot allocate memory)


### What is the sequence quality like? 

I used the FastQC program from the command line to get an idea of what my sequences look like. This isn't incredibly informative because the forward / reverse raw data files are a combination of all reads across all individuals, but it is good to know what, on average, the per base sequence quality looks like. 

FastQC also visualizes: per sequence quality scores, per base sequence content, per base GC content, sequence length distribution, among others. 

[Forward Sequences](file:///D:/Pacific%20cod/DataAnalysis/raw_data/L1_PE150/FastQC/643_S1_L001_R1_001_fastqc.html)

[Reverse Sequences](file:///D:/Pacific%20cod/DataAnalysis/raw_data/L1_PE150/FastQC/643_S1_L001_R2_001_fastqc.html)

**Trends in per-base sequence quality**

1. Forward Sequence, decrease in quality from 7-15bp: This is the restriction site. Low quality likely a result of low diversity of nucleotide bases in this region
2. Forward Sequence, decrease in quality toward end of sequence: quality tends to decrease further along the sequence (mentioned in book). 
3. Reverse Sequence, overall low quality: In my lab's paired-end RADseq data, it tends to be a trend that the reverse reads are lower quality overall than the forward reads. Not sure about the source of this, but I may have to do another lane of paired-end sequencing to get better reverse reads; I want higher quality since I'll need to align them to the Atlantic cod genome